### ISSUE TITLE
Problems with caching on self-hosted runners

### ISSUE THREAD

**Issue by MPV** [scala-steward-org/scala-steward-action#52]
As mentioned in https://github.com/scala-steward-org/scala-steward-action/pull/38#issuecomment-637417913, Scala Steward seems to do some kind of caching (that maybe isn't useful for ephemeral GitHub Action runners), opening this issue for that.

**Comment by MPV** [comment]
A possible workaround might be to add a GHA step that removes all/some of those files...?

**Comment by bpg** [comment]
Caching is an interesting topic that I'd like to discuss a bit more. I think there are some cases where you'd like to utilize results from previous action runs, for example, a big project with lots of dependencies and/or lots of open PRs from Scala Steward, or an action that handles updates for multiple projects.

Running every time from scratch will eat up into your action minutes allotment. Also, Scala Steward has some features that used the persisted (in the wordspace) state, like scanning frequency, et. al. I was really thinking into adding a caching step to my own projects just to deal with all of this.

So, if we're talking about adding a behaviour in the action to _wipe out the workspace_ before calling SS, I would suggest to 
- add a configuration input to enable/disable this behaviour
- have it enabled by default (as I assume the majority of action users are small single project repos)
- explain all of this in documentation and add configuration examples with explicit workspace caching

Thoughts?

**Comment by MPV** [comment]
The current/built-in caching usage of Scala Steward is only per runner and that cache might not be as helpful as compared to how one might otherwise run Scala Steward standalone (towards multiple repos). This in contrast with this action, where the default is to just run towards the same repo.

I’ll see if I can troubleshoot what issues I’ve gotten that have been cached (preventing a rerun to succeed). Essentially whether any caching is performed could be less of an issue if I could rerun and get it to retry creating PRs etc (when it fails on an that is resolved by just retrying).

**Comment by MPV** [comment]
Leaving a note to remember: 
If I (or anyone else) runs into this again, we should probable check the contents of that file (from output above):
```
./scala-steward/workspace/store/refresh_error/v1/github/my-org/my-repo/refresh_error.json
```

Another workaround could be to just remove that file (until we've found a suitable way forward regarding caching per runner/repo/etc).

**Comment by MPV** [comment]
`refresh_error` aside...

Do we have any suggestions on how one might do caching using the [`actions/cache`](https://github.com/marketplace/actions/cache) cache? 
(to cache Scala Steward workspace files between different runners, for the same repo)

Had you given this a try @bpg?

**Comment by MPV** [comment]
To share, here's what I'm trying at the moment:

```diff
diff --git a/.github/workflows/scala-steward.yml b/.github/workflows/scala-steward.yml
index 9c79225..d8cbe8f 100644
--- a/.github/workflows/scala-steward.yml
+++ b/.github/workflows/scala-steward.yml
@@ -1,25 +1,29 @@
 name: Scala Steward

 on:
   # This workflow will launch at 00:00 every Sunday
   schedule:
     - cron: '0 0 * * 0'
   repository_dispatch:
     types: [scala-steward]

 jobs:
   scala-steward:
     runs-on: self-hosted
     name: Launch Scala Steward
     steps:
+      - run: rm -rf ~/scala-steward
+      - uses: actions/cache@v2
+        with:
+          path: ~/scala-steward
+          key: ${{ runner.os }}
       - name: Setup Java and Scala
         uses: olafurpg/setup-scala@v5
       - name: Launch Scala Steward
         uses: scala-steward-org/scala-steward-action@v2
         with:
           github-token: ${{ secrets.ORG_LEVEL_GITHUB_TOKEN }}
           author-email: org-level-robot-user@my-org.com
           author-name: org-level-robot-user
-      - run: rm -rf ~/scala-steward
```

I get these results when running the above:

Before scala-steward-action:
```
Run actions/cache@v2
  with:
    path: ~/scala-steward
    key: Linux
Cache not found for input keys: Linux
```

After:
```
Post Run actions/cache@v2
Cache saved successfully
Post job cleanup.
/bin/tar -z -cf cache.tgz -P -C /home/runner/_work/my-repo/my-repo --files-from manifest.txt
Cache saved successfully
```
🎉 

And in the next job:
```
Run actions/cache@v2
Cache Size: ~0 MB (5867 B)
/bin/tar -z -xf /home/runner/_work/_temp/1a7847a1-58fa-482c-9b00-0bcf25638667/cache.tgz -P -C /home/runner/_work/my-repo/my-repo
Cache restored from key: Linux
```
+
```
Post Run actions/cache@v2
Post job cleanup.
Cache hit occurred on the primary key Linux, not saving cache.
```
😭

**Comment by MPV** [comment]
Tried with caching based on `*.sbt` files:

```diff
diff --git a/.github/workflows/scala-steward.yml b/.github/workflows/scala-steward.yml
index d8cbe8f..43c2249 100644
--- a/.github/workflows/scala-steward.yml
+++ b/.github/workflows/scala-steward.yml
@@ -18,7 +18,7 @@ jobs:
       - uses: actions/cache@v2
         with:
           path: ~/scala-steward
-          key: ${{ runner.os }}
+          key: ${{ runner.os }}-sbt-${{ hashFiles('**/*.sbt') }}
       - name: Setup Java and Scala
         uses: olafurpg/setup-scala@v5
       - name: Launch Scala Steward
```

Unfortunately, that just resolved into `Linux-sbt-` (no hash), since I had forgotten to checkout the repo (have any `actions/checkout` actions).

I added that now, and get:
```
Run actions/cache@v2
Cache Size: ~0 MB (8052 B)
/bin/tar -z -xf /home/runner/_work/_temp/4d38ab15-b49a-4a1c-ab30-7ce433cc8161/cache.tgz -P -C /home/runner/_work/sbt-services/sbt-services
Cache restored from key: Linux-sbt-b19d6d84297c3469e702a9fdf501c41235ffe8bd4b62aed495a8ca62cd1b8fd0
```
+
```
Post Run actions/cache@v2
Post job cleanup.
Cache hit occurred on the primary key Linux-sbt-b19d6d84297c3469e702a9fdf501c41235ffe8bd4b62aed495a8ca62cd1b8fd0, not saving cache.
```

However, one problem here is that even though a dependency might have changed, the cache isn't updated. Changes to my `*.sbt` files needs to happen before the cache is saved into again — note the `not saving cache` above.

**Comment by MPV** [comment]
Also, note how the cache is expanded into my workspace for GitHub Actions:
(`-C /home/runner/_work/my-repo/my-repo` and not `~/scala-steward`)

```
Run actions/cache@v2
  with:
    path: ~/scala-steward
    key: Linux-sbt-b19d6d84297c3469e702a9fdf501c41235ffe8bd4b62aed495a8ca62cd1b8fd0
Cache Size: ~0 MB (8052 B)
/bin/tar -z -xf /home/runner/_work/_temp/17139ec5-0305-4040-9307-8e15c1de1709/cache.tgz -P -C /home/runner/_work/my-repo/my-repo
Cache restored from key: Linux-sbt-b19d6d84297c3469e702a9fdf501c41235ffe8bd4b62aed495a8ca62cd1b8fd0
```

**Comment by MPV** [comment]
Maybe we should revisit the changes made in https://github.com/scala-steward-org/scala-steward-action/pull/42 (where we moved the workspace into ~/scala-workspace)...

See:
https://github.com/scala-steward-org/scala-steward-action/blob/3199096c4fd404510a0849275ac40aacf67d762d/src/files.ts#L26 

...and maybe move the Scala Steward workspace into the GitHub Actions workspace for the repo instead @alejandrohdezma (so caching can be done per repo, and not per runner)?

**Comment by alejandrohdezma** [comment]
Hey! So sorry, I don't know why but I totally miss this issue/conversation. Thank you both so much for all this investigation, this is indeed really interesting!

I did some testing with cache on one of the alpha versions using [toolkit/cache](https://github.com/actions/toolkit/tree/master/packages/cache) but I remember I found some errors and postpone it.

One first thing we could do is using the [post](https://help.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#post) action to cleanup the directory so people don't find this problem on self runners.

**Comment by alejandrohdezma** [comment]
Also, @fthomas could you tell us a bit about which things should be cached from the Scala Steward workspace? That way I can do some testing with it in a different branch :)

**Comment by bpg** [comment]
> Do we have any suggestions on how one might do caching using the [actions/cache](https://github.com/marketplace/actions/cache) action?
> 
> (to cache Scala Steward workspace files between different runners/runs for the same repo)
> 
> Had you given this a try @bpg?
I haven't got far, will try again on the weekend.

In my other project I have a CircleCI Scala Steward custom job running on multiple projects, and caching is a huge help there, taking build time from 6..8 minutes to tens of seconds. So I think it is important to have an option to keep SS cache at the end, and reuse it next time, one way or another.

**Comment by fthomas** [comment]
I would cache everything of the Scala Steward workspace except the `store/refresh_error` directory. The `refresh_error` store is used to temporarily ignore repos whose build can't be loaded by Scala Steward. This is useful when Scala Steward is working on a lot of repos and should not be slowed down by a few repos that have broken builds. I guess it is less useful if Scala Steward runs as GH action where it isn't desirable to ignore broken builds because the operator of the action is able to fix the builds or the action.

**Comment by alejandrohdezma** [comment]
@MPV I have created #57 to address the problem of files created by this action being left behind. Let me know if that would be enough :)

**Comment by alejandrohdezma** [comment]
Thank you very much @fthomas for that explanation! 😸 

I'll try to address that in a new PR today

**Comment by MPV** [comment]
Or, could we change it like this?

1. The action (or Steward) has it's workspace within the repo workspace. 
2. The action (or Steward) cleans any `refresh_error` files/directories after each run. 

Then we could benefit from cross-runners caching (1 + `actions/cache`) and still no get blocked runs due to refresh error locks (2).

**Comment by alejandrohdezma** [comment]
Either way, I think we should remove all the files created by the action since it is what's encouraged by Github so it won't matter if the workspace is in the home directory or the repo workspace and it will be fixed by using `action/cache` in any case

**Comment by MPV** [comment]
> Either way, I think we should remove all the files created by the action since it is what's encouraged by Github so it won't matter if the workspace is in the home directory or the repo workspace and it will be fixed by using `action/cache` in any case

Could you share a link/example to “encouraged by GitHub”. I searched but couldn’t find that, for example here: https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#filesystems-on-github-hosted-runners

Also, were you able to use `action/cache` to cache anything outside the repo workspace?

**Comment by fthomas** [comment]
Right, the `repos` directory does not need to be cached.

**Comment by alejandrohdezma** [comment]
okey dokey! Thanks @fthomas

**Comment by MPV** [comment]
@alejandrohdezma To add another viewpoint to this: 
If we enable caching as in #59, how long will dependencies be cached for?

If dependencies are cached for X days, then it won't be much value in running this action more often than every X days, no?

I have a use-case where we'd like to try using Scala Steward to create fast and semi-automatic bumping of downstream versions from an upstream repo. 

Would setting a TTL for the cache be a suitable solution for this?

For example, I could have: 
- a `weekly` workflow with longer TTL for generic dependencies
- a `repository_dispatch` workflow with longer TTL for being triggered by changes by upstream repos

Alternatively, if there were a setting for Scala Steward itself to drop cache for a specific dependency and force-update just that one (similar to what I suggested in https://github.com/fthomas/scala-steward/issues/1470).

**Comment by alejandrohdezma** [comment]
> @alejandrohdezma To add another viewpoint to this:
> If we enable caching as in #59, how long will dependencies be cached for?
> 
> If dependencies are cached for X days, then it won't be much value in running this action more often than every X days, no?

[Caches are stored for a maximum of 7 days](https://github.com/actions/toolkit/tree/master/packages/cache), so yes, it won't have any effect if the action is run more often than that.

> I have a use-case where we'd like to try using Scala Steward to create fast and semi-automatic bumping of downstream versions from an upstream repo.
> 
> Would setting a TTL for the cache be a suitable solution for this?

As far as I know, setting TTL for actions/cache is not allowed 😿 

> Alternatively, if there were a setting for Scala Steward itself (@fthomas) to drop cache for a specific dependency and force-update just that one (similar to what I suggested in [fthomas/scala-steward#1470](https://github.com/fthomas/scala-steward/issues/1470)).

If this is enabled in scala-steward it shouldn't be to hard to add it to the action :)

**Comment by fthomas** [comment]
Scala Steward already has a `--cache-ttl` option that controls how often it checks for new versions. The default value is `2hours`.

**Comment by MPV** [comment]
> Scala Steward already has a `--cache-ttl` option that controls how often it checks for new versions. The default value is `2hours`.

Opened #62 to be able to set custom cache TTL.
